src/libPMacc/include/nvidia/reduce/Reduce.hpp

```diff
@@ -57,38 +56,39 @@ namespace PMacc
                     /* create a pointer with the right type*/
                     Type* s_mem=(Type*)s_mem_extern;
 
-                    if (tid >= src_count) return; /*end not needed threads*/
+                    if (tid >= src_count)
+                        return; /*end not needed threads*/
 
                     /*fill shared mem*/
                     Type r_value = src[tid];
-                    /*reduce not readed global memory to shared*/
+                    /*reduce not read global memory to shared*/
                     uint32_t i = tid + globalThreadCount;
                     while (i < src_count)
                     {
                         func(r_value, src[i]);
                         i += globalThreadCount;
                     }
-                    s_mem[l_tid] = r_value;
+                    s_mem[threadIdx.x] = r_value;
                     __syncthreads();
                     /*now reduce shared memory*/
                     uint32_t chunk_count = blockDim.x;
-                    uint32_t active_threads;
 
                     while (chunk_count != 1)
                     {
-                        const float half_threads = (float) chunk_count / 2.0f;
-                        active_threads = float2uint(half_threads);
-                        if (threadIdx.x != 0 && l_tid >= active_threads) return; /*end not needed threads*/
-
+                        /* Half number of chunks (rounded down) */
+                        uint32_t active_threads = chunk_count / 2;
+                        if (threadIdx.x >= active_threads)
+                            return; /*end not needed threads*/
 
-                        chunk_count = ceilf(half_threads);
-                        func(s_mem[l_tid], s_mem[l_tid + chunk_count]);
+                        /* New chunks is half number of chunks rounded up for uneven counts
+                         * --> threadIdx.x=0 will reduce the single element for an odd number of values at the end */
+                        chunk_count = (chunk_count + 1) / 2;
```

Comment:
This is not the same like `ceil(chunk_count/2.)` it should be `chunk_count = (chunk_count + chunk_size  - 1 ) / 2;` to be equal.